# 使用`print()`

# 断言
```python
def foo(s):
    n = int(s)
    assert n != 0, 'n is zero!' # 如果断言失败，assert语句本身就会抛出AssertionError
    return 10 / n

def main():
    foo('0')

$ python err.py
Traceback (most recent call last):
  ...
AssertionError: n is zero!
```
启动python解释器时，可以通过 `-O`来关闭断言
```python -O err.py```

# logging
把print()替换为logging是第3种方式，和assert比，logging不会抛出错误，而且可以输出到文件：
```python
# logging可以指定信息级别
# 有debug，info，warning，error等几个级别，当我们指定level=INFO时，logging.debug就不起作用了。同理，指定level=WARNING后，debug和info就不起作用了。这样一来，你可以放心地输出不同级别的信息，也不用删除，最后统一控制输出哪个级别的信息。
# 可以统一配置
import logging
logging.basicConfig(level=logging.INFO)

logging.info('n = %d' % n)
```

# pdb
```python
# err.py
s = '0'
n = int(s)
print(10 / n)

# 以参数-m pdb启动后，pdb定位到下一步要执行的代码-> s = '0'。输入命令l来查看代码：
$ python -m pdb err.py
# 报错 按 l
(Pdb) l
  1     # err.py
  2  -> s = '0'
  3     n = int(s)
  4     print(10 / n)

# 输入命令n可以单步执行代码：
(Pdb) n
# 任何时候都可以输入命令p 变量名来查看变量：
(Pdb) p s # '0'
# 输入命令q结束调试，退出程序：
(Pdb) q

# 缺点就是麻烦
```

# pdb.set_trace()
这个方法也是用pdb,不过不需要但不执行，就类似断点执行
```python
# err.py
import pdb

s = '0'
n = int(s)
pdb.set_trace() # 运行到这里会自动暂停
print(10 / n)

# 运行代码，程序会自动在pdb.set_trace()暂停并进入pdb调试环境，可以用命令p查看变量，或者用命令c继续运行
# 也麻烦
```

# 使用IDE